File: Problems/Linked-lists/Reverse-Linked-List/Reverse-Linked-List.cs
using System;

// Using recursion - Time: O(n)

public class ListNode
{
    public int val;
    public ListNode next;
    public ListNode(int val = 0, ListNode next = null)
    {
        this.val = val;
        this.next = next;
    }
}

public class Solution
{
    private (ListNode, ListNode) Reverse(ListNode head)
    {
        if (head.next == null) return (head, head);
        
        var (h, t) = Reverse(head.next);
        t.next = head;
        head.next = null;

        return (h, head);
    }

    public ListNode ReverseList(ListNode head)
    {
        if (head == null) return null;
        return Reverse(head).Item1;
    }
}

// Using iteration - Time: O(n)

// public class Solution {
//     public ListNode ReverseList(ListNode head) {
//         ListNode h = new ListNode();
//         while (head != null) {
//             var p = head;
//             head = head.next;
//             p.next = h.next;
//             h.next = p;
//         }
//         return h.next;
//     }
// }

class Program
{
    static void PrintList(ListNode node)
    {
        Console.Write("[");
        while (node != null)
        {
            Console.Write(node.val);
            node = node.next;
            if (node != null)
            {
                Console.Write(", ");
            }
        }
        Console.Write("]");
    }
    static void Main(string[] args)
    {
        ListNode list = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));

        Console.Write("Input: head = ");
        PrintList(list);
        Console.WriteLine();

        Solution sol = new Solution();
        ListNode reversedList = sol.ReverseList(list);

        Console.Write("Output: ");
        PrintList(reversedList);
        Console.WriteLine();
    }
}

File: Problems/Linked-lists/Reverse-Linked-List/Reverse-Linked-List.cpp
#include <iostream>
#include <array>

// Using recursion - Time: O(n)

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
    std::array<ListNode*, 2> reverse(ListNode *head) 
    {
        if (!head->next) return {head, head};
    
        auto [h, t] = reverse(head->next);
        t->next = head;
        head->next = NULL;
    
        return {h, head};
    }

public:
    ListNode* reverseList(ListNode* head) 
    {
        if (!head) return NULL;
        return reverse(head)[0];
    }
};

// Using iteration - Time: O(n)

// class Solution {
// public:
//     ListNode* reverseList(ListNode* head) {
//         ListNode h;
//         while (head) {
//             auto p = head;
//             head = head->next;
//             p->next = h.next;
//             h.next = p;
//         }
//         return h.next;
//     }
// };

void printList(ListNode *node)
{
    std::cout << "[";
    while (node != nullptr)
    {
        std::cout << node->val;
        node = node->next;
        if (node != nullptr)
        {
            std::cout << ", ";
        }
    }
    std::cout << "]";
}

int main()
{
    ListNode *list = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));

    std::cout << "Input: head = ";
    printList(list);
    std::cout << std::endl;

    Solution sol;
    ListNode *reversedList = sol.reverseList(list);

    std::cout << "Output: ";
    printList(reversedList);
    std::cout << std::endl;

    delete list;

    return 0;
}

